AuthSAML.php

```diff
@@ -131,125 +151,165 @@ public function beforeLogin() {
+        }
         $ssp = $this->get_saml_instance();
-
+        if (is_null($ssp)) {
+            return;
+        }
         if ($ssp->isAuthenticated()) {
-
             $sUser = $this->getUserName();
             $name = $this->getUserCommonName();
             $mail = $this->getUserMail();
 
             $oUser = $this->api->getUserByName($sUser);
-
             $auto_create_users = $this->get('auto_create_users', null, null, true);
 
             if (is_null($oUser) and $auto_create_users) {
-
                 // Create new user
-                $oUser = new User;
+                $oUser = new User();
                 $oUser->users_name = $sUser;
                 $oUser->setPassword(createPassword());
                 $oUser->full_name = $name;
                 $oUser->parent_id = 1;
                 $oUser->email = $mail;
 
                 if ($oUser->save()) {
-                    $permission = new Permission;
-
+                    $permission = new Permission();
                     Permission::model()->setGlobalPermission($oUser->uid, 'auth_saml');
-
                     $oUser = $this->api->getUserByName($sUser);
-
                     $this->pluginManager->dispatchEvent(new PluginEvent('newUserLogin', $this));
-
                     $this->setAuthSuccess($oUser);
                 } else {
+                    $this->subscribe('afterFailedLoginAttempt');
                     $this->setAuthFailure(self::ERROR_USERNAME_INVALID);
                 }
             } elseif (is_null($oUser)) {
-                throw new CHttpException(401, gT("We are sorry but you do not have an account."));
+                $this->subscribe('afterFailedLoginAttempt');
+                $this->setAuthFailure(self::ERROR_USERNAME_INVALID);
+            } elseif (!Permission::model()->hasGlobalPermission('auth_saml', 'read', $oUser->uid)) {
+                $this->subscribe('afterFailedLoginAttempt');
+                $this->setAuthFailure(self::ERROR_AUTH_METHOD_INVALID, $this->gT('Invalid SAML account for this service'));
             } else {
-
+                /* @todo Check if user have still access (expire and active) */
                 // *** Update user ***
                 $auto_update_users = $this->get('auto_update_users', null, null, true);
-
                 if ($auto_update_users) {
                     $changes = array (
                         'full_name' => $name,
                         'email' => $mail,
                     );
-
                     User::model()->updateByPk($oUser->uid, $changes);
                     $oUser = $this->api->getUserByName($sUser);
                 }
-
                 $this->setAuthSuccess($oUser);
             }
         }
         $flag = $this->get('simplesamlphp_cookie_session_storage', null, null, true);
-        if ($flag){
-            $session = SimpleSAML_Session::getSessionFromRequest();
+        if ($flag) {
+            $session = \SimpleSAML\Session::getSessionFromRequest();
```

Comment:
This line causes the following error:
` PHP Fatal error:  Cannot declare class SimpleSAML_Session, because the name is already in use in /usr/share/simplesamlphp/lib/SimpleSAML/Session.php on line 19`
If i revert this line to previous version, it works ok.

Reply:
There are no SimpleSAML_Session Class

What version of simplesamlphp did you test ? Maybe it's related ?

I use last version (


```
Class "SimpleSAML_Session" not found

/home/limesurvey.shnoulle.net/Limesurvey-SAML-Authentication/AuthSAML.php(269)

257                     $changes = array (
258                         'full_name' => $name,
259                         'email' => $mail,
260                     );
261                     User::model()->updateByPk($oUser->uid, $changes);
262                     $oUser = $this->api->getUserByName($sUser);
263                 }
264                 $this->setAuthSuccess($oUser);
265             }
266         }
267         $flag = $this->get('simplesamlphp_cookie_session_storage', null, null, true);
268         if ($flag) {
269             $session = SimpleSAML_Session::getSessionFromRequest();
270             //~ $session = \SimpleSAML\Session::getSessionFromRequest();
271             $session->cleanup();
272         }
273     }
274 
275     /**
276      * Initialize SAML authentication
277      * @return null|\SimpleSAML
278      */
279     public function get_saml_instance()
280     {
281         if ($this->ssp == null) {

```

![Uploading Capture d’écran du 2024-03-04 11-11-56.png…]()


Maybe we can check if class exist depend on simplesaml version ?

Reply:
We are on 1.15, very likely this is the reason.

Reply:
Please test with https://github.com/auth-it-center/Limesurvey-SAML-Authentication/pull/6/commits/81458de1fec13f12ad9f690084a3c6ef8f19ffb1

Reply:
> We are on 1.15, very likely this is the reason.

2.1.3 for me

Have issue with twig with this version